// graphene-pk11/build/objects/certs/cert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const pkcs11 = tslib_1.__importStar(require("pkcs11js"));
const storage_1 = require("../storage");
var CertificateType;
(function (CertificateType) {
    CertificateType[CertificateType["X_509"] = pkcs11.CKC_X_509] = "X_509";
    CertificateType[CertificateType["X_509_ATTR_CERT"] = pkcs11.CKC_X_509_ATTR_CERT] = "X_509_ATTR_CERT";
    CertificateType[CertificateType["WTLS"] = pkcs11.CKC_WTLS] = "WTLS";
})(CertificateType = exports.CertificateType || (exports.CertificateType = {}));
var CertificateCategory;
(function (CertificateCategory) {
    CertificateCategory[CertificateCategory["Unspecified"] = 0] = "Unspecified";
    CertificateCategory[CertificateCategory["TokenUser"] = 1] = "TokenUser";
    CertificateCategory[CertificateCategory["Authority"] = 2] = "Authority";
    CertificateCategory[CertificateCategory["OtherEntity"] = 3] = "OtherEntity";
})(CertificateCategory = exports.CertificateCategory || (exports.CertificateCategory = {}));
class Certificate extends storage_1.Storage {
    get type() {
        return this.get("certType");
    }
    set type(v) {
        this.set("certType", v);
    }
    get trusted() {
        return this.get("trusted");
    }
    set trusted(v) {
        this.set("trusted", v);
    }
    get category() {
        return this.get("certCategory");
    }
    set category(v) {
        this.set("certCategory", v);
    }
    get checkValue() {
        return this.get("checkValue");
    }
    set checkValue(v) {
        this.set("checkValue", v);
    }
    get startDate() {
        return this.get("startDate");
    }
    set startDate(v) {
        this.set("startDate", v);
    }
    get endDate() {
        return this.get("endDate");
    }
    set endDate(v) {
        this.set("endDate", v);
    }
}
exports.Certificate = Certificate;
