// graphene-pk11/build/objects/certs/wtls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cert_1 = require("./cert");
class WtlsCertificate extends cert_1.Certificate {
    get subject() {
        return this.get("subject");
    }
    set subject(v) {
        this.set("subject", v);
    }
    get issuer() {
        return this.get("issuer");
    }
    set issuer(v) {
        this.set("issuer", v);
    }
    get id() {
        return this.get("id");
    }
    set id(v) {
        this.set("id", v);
    }
    get value() {
        return this.get("value");
    }
    set value(v) {
        this.set("value", v);
    }
    get url() {
        return this.get("url");
    }
    set url(v) {
        this.set("url", v);
    }
    get serialNumber() {
        return this.get("serial");
    }
    set serialNumber(v) {
        this.set("serial", v);
    }
    get subjectKeyIdentifier() {
        return this.get("ski");
    }
    set subjectKeyIdentifier(v) {
        this.set("ski", v);
    }
    get authorityKeyIdentifier() {
        return this.get("aki");
    }
    set authorityKeyIdentifier(v) {
        this.set("aki", v);
    }
}
exports.WtlsCertificate = WtlsCertificate;
