// graphene-pk11/build/objects/certs/x509.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cert_1 = require("./cert");
var JavaMIDP;
(function (JavaMIDP) {
    JavaMIDP[JavaMIDP["Unspecified"] = 0] = "Unspecified";
    JavaMIDP[JavaMIDP["Manufacturer"] = 1] = "Manufacturer";
    JavaMIDP[JavaMIDP["Operator"] = 2] = "Operator";
    JavaMIDP[JavaMIDP["ThirdParty"] = 3] = "ThirdParty";
})(JavaMIDP = exports.JavaMIDP || (exports.JavaMIDP = {}));
class X509Certificate extends cert_1.Certificate {
    get subject() {
        return this.get("subject");
    }
    set subject(v) {
        this.set("subject", v);
    }
    get id() {
        return this.get("id");
    }
    set id(v) {
        this.set("id", v);
    }
    get issuer() {
        return this.get("issuer");
    }
    set issuer(v) {
        this.set("issuer", v);
    }
    get serialNumber() {
        return this.get("serial").toString("hex");
    }
    set serialNumber(v) {
        this.set("serial", Buffer.from(v, "hex"));
    }
    get value() {
        return this.get("value");
    }
    set value(v) {
        this.set("value", v);
    }
    get url() {
        return this.get("url");
    }
    set url(v) {
        this.set("url", v);
    }
    get subjectKeyIdentifier() {
        return this.get("ski");
    }
    set subjectKeyIdentifier(v) {
        this.set("ski", v);
    }
    get authorityKeyIdentifier() {
        return this.get("aki");
    }
    set authorityKeyIdentifier(v) {
        this.set("aki", v);
    }
    get java() {
        return this.get("javaDomain");
    }
    set java(v) {
        this.set("javaDomain", v);
    }
}
exports.X509Certificate = X509Certificate;
