// node-webcrypto-p11/build/key.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core = require("webcrypto-core");
const graphene_pk11_1 = require("graphene-pk11");
class CryptoKey extends core.CryptoKey {
    constructor(key, alg) {
        super();
        this.usages = [];
        this.p11Object = key;
        switch (key.class) {
            case graphene_pk11_1.ObjectClass.PUBLIC_KEY:
                this.initPublicKey(key.toType());
                break;
            case graphene_pk11_1.ObjectClass.PRIVATE_KEY:
                this.initPrivateKey(key.toType());
                break;
            case graphene_pk11_1.ObjectClass.SECRET_KEY:
                this.initSecretKey(key.toType());
                break;
            default:
                throw new core.CryptoError(`Wrong incoming session object '${graphene_pk11_1.ObjectClass[key.class]}'`);
        }
        this.algorithm = alg;
        this.id = CryptoKey.getID(key);
        this.onAssign();
    }
    static getID(p11Key) {
        let name;
        switch (p11Key.class) {
            case graphene_pk11_1.ObjectClass.PRIVATE_KEY:
                name = "private";
                break;
            case graphene_pk11_1.ObjectClass.PUBLIC_KEY:
                name = "public";
                break;
            case graphene_pk11_1.ObjectClass.SECRET_KEY:
                name = "secret";
                break;
            default:
                throw new Error(`Unsupported Object type '${graphene_pk11_1.ObjectClass[p11Key.class]}'`);
        }
        return `${name}-${p11Key.handle.toString("hex")}-${p11Key.id.toString("hex")}`;
    }
    get key() {
        return this.p11Object.toType();
    }
    toJSON() {
        return {
            algorithm: this.algorithm,
            type: this.type,
            usages: this.usages,
            extractable: this.extractable,
        };
    }
    initPrivateKey(key) {
        this.p11Object = key;
        this.type = "private";
        try {
            this.extractable = key.extractable;
        }
        catch (e) {
            this.extractable = false;
        }
        this.usages = [];
        if (key.decrypt) {
            this.usages.push("decrypt");
        }
        if (key.derive) {
            this.usages.push("deriveKey");
            this.usages.push("deriveBits");
        }
        if (key.sign) {
            this.usages.push("sign");
        }
        if (key.unwrap) {
            this.usages.push("unwrapKey");
        }
    }
    initPublicKey(key) {
        this.p11Object = key;
        this.type = "public";
        this.extractable = true;
        if (key.encrypt) {
            this.usages.push("encrypt");
        }
        if (key.verify) {
            this.usages.push("verify");
        }
        if (key.wrap) {
            this.usages.push("wrapKey");
        }
    }
    initSecretKey(key) {
        this.p11Object = key;
        this.type = "secret";
        try {
            this.extractable = key.extractable;
        }
        catch (e) {
            this.extractable = false;
        }
        if (key.sign) {
            this.usages.push("sign");
        }
        if (key.verify) {
            this.usages.push("verify");
        }
        if (key.encrypt) {
            this.usages.push("encrypt");
        }
        if (key.decrypt) {
            this.usages.push("decrypt");
        }
        if (key.wrap) {
            this.usages.push("wrapKey");
        }
        if (key.unwrap) {
            this.usages.push("unwrapKey");
        }
        if (key.derive) {
            this.usages.push("deriveKey");
            this.usages.push("deriveBits");
        }
    }
    onAssign() {
    }
}
exports.CryptoKey = CryptoKey;
