// graphene-pk11/build/objects/keys/private.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const key_1 = require("./key");
class PrivateKey extends key_1.Key {
    get subject() {
        return this.get("subject");
    }
    set subject(v) {
        this.set("subject", v);
    }
    get sensitive() {
        return this.get("sensitive");
    }
    set sensitive(v) {
        this.set("sensitive", v);
    }
    get decrypt() {
        try {
            return this.get("decrypt");
        }
        catch (err) {
            return false;
        }
    }
    set decrypt(v) {
        this.set("decrypt", v);
    }
    get sign() {
        try {
            return this.get("sign");
        }
        catch (err) {
            return false;
        }
    }
    set sign(v) {
        this.set("sign", v);
    }
    get signRecover() {
        try {
            return this.get("signRecover");
        }
        catch (err) {
            return false;
        }
    }
    set signRecover(v) {
        this.set("signRecover", v);
    }
    get unwrap() {
        try {
            return this.get("unwrap");
        }
        catch (err) {
            return false;
        }
    }
    set unwrap(v) {
        this.set("unwrap", v);
    }
    get extractable() {
        return this.get("extractable");
    }
    set extractable(v) {
        this.set("extractable", v);
    }
    get alwaysSensitive() {
        return this.get("alwaysSensitive   ");
    }
    set alwaysSensitive(v) {
        this.set("alwaysSensitive", v);
    }
    get neverExtractable() {
        return this.get("neverExtractable");
    }
    set neverExtractable(v) {
        this.set("neverExtractable", v);
    }
    get wrapTrusted() {
        return this.get("wrapWithTrusted");
    }
    set wrapTrusted(v) {
        this.set("wrapWithTrusted", v);
    }
    get template() {
        throw new Error("Not implemented");
    }
    set template(v) {
        throw new Error("Not implemented");
    }
    get alwaysAuthenticate() {
        return this.get("alwaysAuth");
    }
    set alwaysAuthenticate(v) {
        this.set("alwaysAuth", v);
    }
}
exports.PrivateKey = PrivateKey;
