// graphene-pk11/build/objects/keys/public.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const key_1 = require("./key");
class PublicKey extends key_1.Key {
    get subject() {
        return this.get("subject");
    }
    set subject(v) {
        this.set("subject", v);
    }
    get encrypt() {
        try {
            return this.get("encrypt");
        }
        catch (err) {
            return false;
        }
    }
    set encrypt(v) {
        this.set("encrypt", v);
    }
    get verify() {
        try {
            return this.get("verify");
        }
        catch (err) {
            return false;
        }
    }
    set verify(v) {
        this.set("verify", v);
    }
    get verifyRecover() {
        try {
            return this.get("verifyRecover");
        }
        catch (err) {
            return false;
        }
    }
    set verifyRecover(v) {
        this.set("verifyRecover", v);
    }
    get wrap() {
        try {
            return this.get("wrap");
        }
        catch (err) {
            return false;
        }
    }
    set wrap(v) {
        this.set("wrap", v);
    }
    get trusted() {
        return this.get("trusted");
    }
    set trusted(v) {
        this.set("trusted", v);
    }
    get template() {
        throw new Error("Not implemented");
    }
    set template(v) {
        throw new Error("Not implemented");
    }
}
exports.PublicKey = PublicKey;
